README.md
# RuCaptcha

[![Gem Version](https://badge.fury.io/rb/rucaptcha.svg)](https://badge.fury.io/rb/rucaptcha)
[![build](https://github.com/huacnlee/rucaptcha/workflows/build/badge.svg)](https://github.com/huacnlee/rucaptcha/actions?query=workflow%3Abuild)

This is a Captcha gem for Rails Applications which generates captcha image by C code.

> NOTE: According to the use of Ruby China, the verification code looks like has a lower than 5% probability of being parsed by OCR and the verification code is cracked (All Image Captcha libs are has same problem). It is recommended that you use the IP rate limit to enhance the protection.
> NOTE: 以 Ruby China 的使用来看，验证码似乎有低于 5% 的概率被 OCR 读取解析 (图片验证码都有这个问题) 导致验证码被破解（我们从日志分析绝大多数是成功的，但偶尔一个成功，配合大量机器攻击，导致注册了很多的垃圾账号），建议你额外配合 IP 频率限制的功能来加强保护。

> 如果你需要更高强度的验证，建议选择商用服务。

[中文介绍和使用说明](https://ruby-china.org/topics/27832)

## Example

<img src="https://user-images.githubusercontent.com/5518/49985459-f8492f80-ffa6-11e8-9ef5-8f8f522e4707.png" width="579px" />

## Feature

- No dependencies. No ImageMagick. No RMagick;
- For Rails Application;
- Simple, Easy to use;
- High performance.

## Usage

Put rucaptcha in your `Gemfile`:

```
gem 'rucaptcha'
```

Create `config/initializers/rucaptcha.rb`

```rb
RuCaptcha.configure do
  # Color style, default: :colorful, allows: [:colorful, :black_white]
  # self.style = :colorful
  # Custom captcha code expire time if you need, default: 2 minutes
  # self.expires_in = 120
  # [Requirement / 重要]
  # Store Captcha code where, this config more like Rails config.cache_store
  # default: Read config info from `Rails.application.config.cache_store`
  # But RuCaptcha requirements cache_store not in [:null_store, :memory_store, :file_store]
  # 默认：会从 Rails 配置的 cache_store 里面读取相同的配置信息，并尝试用可以运行的方式，用于存储验证码字符
  # 但如果是 [:null_store, :memory_store, :file_store] 之类的，你可以通过下面的配置项单独给 RuCaptcha 配置 cache_store
  self.cache_store = :mem_cache_store
  # 如果想要 disable cache_store 的 warning，就设置为 true，default false
  # self.skip_cache_store_check = true
  # Chars length, default: 5, allows: [3 - 7]
  # self.length = 5
  # enable/disable Strikethrough.
  # self.strikethrough = true
  # enable/disable Outline style
  # self.outline = false
end
```

RuCaptcha 没有使用 Rails Session 来存储验证码信息，因为 Rails 的默认 Session 是存储在 Cookie 里面，如果验证码存在里面会存在 [Replay attack](https://en.wikipedia.org/wiki/Replay_attack) 漏洞，导致验证码关卡被攻破。

所以我在设计上要求 RuCaptcha 得配置一个可以支持分布式的后端存储方案例如：Memcached 或 Redis 以及其他可以支持分布式的 cache_store 方案。

同时，为了保障易用性，默认会尝试使用 `:file_store` 的方式，将验证码存在应用程序的 `tmp/cache/rucaptcha/session` 目录（但请注意，多机器部署这样是无法正常运作的）。

所以，我建议大家使用的时候，配置上 `cache_store` （详见 [Rails Guides 缓存配置部分](https://ruby-china.github.io/rails-guides/caching_with_rails.html#%E9%85%8D%E7%BD%AE)的文档）到一个 Memcached 或 Redis，这才是最佳实践。

#

(RuCaptha do not use Rails Session to store captcha information. As the default session is stored in Cookie in Rails, there's a [Replay attack](https://en.wikipedia.org/wiki/Replay_attack) bug which may causes capthcha being destroyed if we store captcha in Rails Session.

So in my design I require RuCaptcha to configure a distributed backend storage scheme, such as Memcached, Redis or other cache_store schemes which support distribution.

Meanwhile, for the ease of use, RuCapthca would try to use `:file_store` by default and store the capthca in `tmp/cache/rucaptcha/session` directory (kindly note that it's not working if deploy on multiple machine).

For recommendation, configure the `cache_store`（more details on [Rails Guides Configuration of Cache Stores](http://guides.rubyonrails.org/caching_with_rails.html#configuration)） to Memcached or Redis, that would be the best practice.)

#

Controller `app/controller/account_controller.rb`

When you called `verify_rucaptcha?`, it uses value from `params[:_rucaptcha]` to validate.

```rb
class AccountController < ApplicationController
  def create
    @user = User.new(params[:user])
    if verify_rucaptcha?(@user) && @user.save
      redirect_to root_path, notice: 'Sign up successed.'
    else
      render 'account/new'
    end
  end
end

class ForgotPasswordController < ApplicationController
  def create
    # without any args
    if verify_rucaptcha?
      to_send_email
    else
      redirect_to '/forgot-password', alert: 'Invalid captcha code.'
    end
  end
end
```

> TIP: Sometimes you may need to keep last verified captcha code in session on `verify_rucaptcha?` method call, you can use `keep_session: true`. For example: `verify_rucaptcha? @user, keep_session: true`.

View `app/views/account/new.html.erb`

```erb
<form method="POST">
  ...
  <div class="form-group">
    <%= rucaptcha_input_tag(class: 'form-control', placeholder: 'Input Captcha') %>
    <%= rucaptcha_image_tag(alt: 'Captcha') %>
  </div>
  ...

  <div class="form-group">
    <button type="submit" class="btn btn-primary">Submit</button>
  </div>
</form>
```

And if you are using [Devise](https://github.com/plataformatec/devise), you can read this reference to add validation: [RuCaptcha with Devise](https://github.com/huacnlee/rucaptcha/wiki/Working-with-Devise).

### Write your test skip captcha validation

for RSpec

```rb
describe 'sign up and login', type: :feature do
  before do
    allow_any_instance_of(ActionController::Base).to receive(:verify_rucaptcha?).and_return(true)
  end

  it { ... }
end
```

for MiniTest

```rb
class ActionDispatch::IntegrationTest
  def sign_in(user)
    ActionController::Base.any_instance.stubs(:verify_rucaptcha?).returns(true)
    post user_session_path \
         'user[email]'    => user.email,
         'user[password]' => user.password
  end
end
```

### Invalid message without Devise

When you are using this gem without Devise, you may find out that the invalid message is missing.
For this case, use the trick below to add your i18n invalid message manually.

```rb
if verify_rucaptcha?(@user) && @user.save
  do_whatever_you_want
  redirect_to someplace_you_want
else
  # this is the trick
  @user.errors.add(:base, t('rucaptcha.invalid'))
  render :new
end
```


CHANGELOG.md
New release please visit:

https://github.com/huacnlee/rucaptcha/releases

## 2.5.5

- Improved image for thicker interference lines.
- Add more colors.

## 2.5.4

- Fix: rucaptcha input maxlength attribute with config value.

## 2.5.3

- Fix session invalid warning, only for development env;

## 2.5.2

- Fix session.id error with upgrade Rails 6.0.2.1 or Rack 2.0.8 (#84)

## 2.5.1

- Fix invalid module name error. (#78)

## 2.5.0

- Support click captcha image to refresh new one by default.
- Use simple tag helper generate captcha img html, for avoid asset_host (#73).

## 2.4.0

- Add skip_cache_store_check configuration. (#63)
- Fix for generate captcha with relative path, not url. (#58)

## 2.3.2

- Change Yellow and Green colors to Pink and Deep Purple to pass WCAG 2.0's contrast test. (#70)

## 2.3.1

- Fix #67 a y chars will invalid error (only in 2.3.0).

## 2.3.0

- Add `config.outline` for use outline style.
- Reduce colors down to 5 (red, blue, green, yellow and black).

## 2.2.0

- Add option `config.length` for support change number chars. (#57)
- Add option `config.strikethrough` for enable or disable strikethrough. (#57)

## 2.1.3

- Windows support fixed with `send_data` method. (#45)

## 2.1.2

- Do not change captcha when `HEAD /rucaptcha`.

## 2.1.1

- Mount engine use `prepend` method to get high priority in config/routes.rb.

## 2.1.0

- Mount Router by default, not need config now.

  > IMPORTANT: Wen you upgrade this version, you need remove `mount RuCaptcha::Engine` line from your `config/routes.rb`

- Default use [:file_store, 'tmp/cache/rucaptcha/session'] as RuCaptcha.config.cache_store, now it can work without any configurations.

> NOTE: But you still need care about `config.cache_store` to setup on a right way.

## 2.0.3

- Use `ActiveSupport.on_load` to extend ActionController and ActionView.

## 2.0.1

- Fix `/rucaptcha` path issue when `config.action_controller.asset_host` has setup with CDN url.

## 2.0.0

_Break Changes!_

WARNING!: This version have so many break changes!

- Use C ext instead of ImageMagick, now it's no dependencies!
- New captcha style.
- Remove `len`, `font_size`, `cache_limit` config key, no support now.
- Output `GIF` format.


ext/rucaptcha/rucaptcha.c
// http://github.com/ITikhonov/captcha
const int gifsize;
void captcha(unsigned char im[70 * 200], unsigned char l[8], int length, int i_line, int i_filter);
void makegif(unsigned char im[70 * 200], unsigned char gif[gifsize], int style);

#include <unistd.h>
#include <stdint.h>
#include <stdlib.h>
#include <fcntl.h>
#include <string.h>
#include <time.h>
#include <ruby.h>
#include "font.h"
#include "colors.h"

static int8_t *lt[];
const int gifsize = 17646;

void makegif(unsigned char im[70 * 200], unsigned char gif[gifsize], int style)
{
  // tag ; widthxheight ; GCT:0:0:7 ; bgcolor + aspect // GCT
  // Image Separator // left x top // widthxheight // Flags
  // LZW code size
  srand(time(NULL));
  int color_len = (int)sizeof(colors) / sizeof(colors[0]);
  int color_idx = rand() % color_len;
  if (style == 0)
  {
    color_idx = 0;
  }
  memcpy(gif, colors[color_idx], 13 + 48 + 10 + 1);

  int x, y;
  unsigned char *i = im;
  unsigned char *p = gif + 13 + 48 + 10 + 1;
  for (y = 0; y < 70; y++)
  {
    *p++ = 250; // Data length 5*50=250
    for (x = 0; x < 50; x++)
    {
      unsigned char a = i[0] >> 4, b = i[1] >> 4, c = i[2] >> 4, d = i[3] >> 4;

      p[0] = 16 | (a << 5);            // bbb10000
      p[1] = (a >> 3) | 64 | (b << 7); // b10000xb
      p[2] = b >> 1;                   // 0000xbbb
      p[3] = 1 | (c << 1);             // 00xbbbb1
      p[4] = 4 | (d << 3);             // xbbbb100
      i += 4;
      p += 5;
    }
  }

  // Data length // End of LZW (b10001) // Terminator // GIF End
  memcpy(gif + gifsize - 4, "\x01"
                            "\x11"
                            "\x00"
                            ";",
         4);
}

static const int8_t sw[200] = {0, 4, 8, 12, 16, 20, 23, 27, 31, 35, 39, 43, 47, 50, 54, 58, 61, 65, 68, 71, 75, 78, 81, 84, 87, 90, 93, 96, 98, 101, 103, 105, 108, 110, 112, 114, 115, 117, 119, 120, 121, 122, 123, 124, 125, 126, 126, 127, 127, 127, 127, 127, 127, 127, 126, 126, 125, 124, 123, 122, 121, 120, 119, 117, 115, 114, 112, 110, 108, 105, 103, 101, 98, 96, 93, 90, 87, 84, 81, 78, 75, 71, 68, 65, 61, 58, 54, 50, 47, 43, 39, 35, 31, 27, 23, 20, 16, 12, 8, 4, 0, -4, -8, -12, -16, -20, -23, -27, -31, -35, -39, -43, -47, -50, -54, -58, -61, -65, -68, -71, -75, -78, -81, -84, -87, -90, -93, -96, -98, -101, -103, -105, -108, -110, -112, -114, -115, -117, -119, -120, -121, -122, -123, -124, -125, -126, -126, -127, -127, -127, -127, -127, -127, -127, -126, -126, -125, -124, -123, -122, -121, -120, -119, -117, -115, -114, -112, -110, -108, -105, -103, -101, -98, -96, -93, -90, -87, -84, -81, -78, -75, -71, -68, -65, -61, -58, -54, -50, -47, -43, -39, -35, -31, -27, -23, -20, -16, -12, -8, -4};

#define MAX(x, y) ((x > y) ? (x) : (y))

static int letter(int n, int pos, unsigned char im[70 * 200], unsigned char swr[200], uint8_t s1, uint8_t s2)
{
  int8_t *p = lt[n];
  unsigned char *r = im + 200 * 16 + pos;
  unsigned char *i = r;
  int sk1 = s1 + pos;
  int sk2 = s2 + pos;
  int mpos = pos;
  int row = 0;
  for (; *p != -101; p++)
  {
    if (*p < 0)
    {
      if (*p == -100)
      {
        r += 200;
        i = r;
        sk1 = s1 + pos;
        row++;
        continue;
      }
      i += -*p;
      continue;
    }

    if (sk1 >= 200)
      sk1 = sk1 % 200;
    int skew = sw[sk1] / 16;
    sk1 += (swr[pos + i - r] & 0x1) + 1;

    if (sk2 >= 200)
      sk2 = sk2 % 200;
    int skewh = sw[sk2] / 70;
    sk2 += (swr[row] & 0x1);

    unsigned char *x = i + skew * 200 + skewh;
    mpos = MAX(mpos, pos + i - r);

    if ((x - im) < 70 * 200)
      *x = (*p) << 4;
    i++;
  }
  return mpos + 3;
}

#define NDOTS 10

uint32_t dr[NDOTS];

static void line(unsigned char im[70 * 200], unsigned char swr[200], uint8_t s1)
{
  int x;
  int sk1 = s1;
  for (x = 0; x < 199; x++)
  {
    if (sk1 >= 200)
      sk1 = sk1 % 200;
    int skew = sw[sk1] / 20;
    sk1 += swr[x] & 0x3 + 1;
    unsigned char *i = im + (200 * (45 + skew) + x);
    i[0] = 0;
    i[1] = 0;
    i[2] = 0;
    i[3] = 0;
    i[4] = 0;
    i[195] = 0;
    i[196] = 0;
    i[197] = 0;
    i[198] = 0;
    i[200] = 0;
  }
}

static void dots(unsigned char im[70 * 200])
{
  int n;
  for (n = 0; n < NDOTS; n++)
  {
    uint32_t v = dr[n];
    unsigned char *i = im + v % (200 * 67);

    i[0] = 0xff;
    i[1] = 0xff;
    i[2] = 0xff;
    i[200] = 0xff;
    i[201] = 0xff;
    i[202] = 0xff;
  }
}

static void blur(unsigned char im[70 * 200])
{
  unsigned char *i = im;
  int x, y;
  for (y = 0; y < 68; y++)
  {
    for (x = 0; x < 198; x++)
    {
      unsigned int c11 = *i, c12 = i[1], c21 = i[200], c22 = i[201];
      *i++ = ((c11 + c12 + c21 + c22) / 4);
    }
  }
}

static void filter(unsigned char im[70 * 200])
{
  unsigned char om[70 * 200];
  unsigned char *i = im;
  unsigned char *o = om;

  memset(om, 0xff, sizeof(om));

  int x, y;
  for (y = 0; y < 70; y++)
  {
    for (x = 4; x < 200 - 4; x++)
    {
      if (i[0] > 0xf0 && i[1] < 0xf0)
      {
        o[0] = 0;
        o[1] = 0;
      }
      else if (i[0] < 0xf0 && i[1] > 0xf0)
      {
        o[0] = 0;
        o[1] = 0;
      }

      i++;
      o++;
    }
  }

  memmove(im, om, sizeof(om));
}

static const char *letters = "abcdafahijklmnopqrstuvwxyz";

void captcha(unsigned char im[70 * 200], unsigned char l[8], int length, int i_line, int i_filter)
{
  unsigned char swr[200];
  uint8_t s1, s2;

  int f = open("/dev/urandom", O_RDONLY);
  read(f, l, 5);
  read(f, swr, 200);
  read(f, dr, sizeof(dr));
  read(f, &s1, 1);
  read(f, &s2, 1);
  close(f);
  memset(im, 0xff, 200 * 70);
  s1 = s1 & 0x7f;
  s2 = s2 & 0x3f;

  int x;
  for (x = 0; x < length; x++)
  {
    l[x] %= 25;
  }
  for (x = length; x < 8; x++)
  {
    l[length] = 0;
  }
  //l[0]%=25; l[1]%=25; l[2]%=25; l[3]%=25; l[4]=0; // l[4]%=25; l[5]=0;
  int p = 30;
  for (x = 0; x < length; x++)
  {
    p = letter(l[x], p, im, swr, s1, s2);
  }

  if (i_line == 1)
  {
    line(im, swr, s1);
  }
  dots(im);
  if (i_filter == 1)
  {
    blur(im);
    filter(im);
  }

  for (x = 0; x < length; x++)
  {
    l[x] = letters[l[x]];
  }
  //l[1]=letters[l[1]]; l[2]=letters[l[2]]; l[3]=letters[l[3]]; //l[4]=letters[l[4]];
}

// #ifdef CAPTCHA
//
// int main() {
//   char l[6];
//   unsigned char im[70*200];
//   unsigned char gif[gifsize];
//
//   captcha(im,l);
//   makegif(im,gif);
//
//   write(1,gif,gifsize);
//   write(2,l,5);
//
//   return 0;
// }
//
// #endif

VALUE RuCaptcha = Qnil;

void Init_rucaptcha();

VALUE create(VALUE self, VALUE style, VALUE length, VALUE line, VALUE filter);

void Init_rucaptcha()
{
  RuCaptcha = rb_define_module("RuCaptcha");
  rb_define_singleton_method(RuCaptcha, "create", create, 4);
}

VALUE create(VALUE self, VALUE style, VALUE length, VALUE line, VALUE filter)
{
  char l[8];
  unsigned char im[80 * 200];
  unsigned char gif[gifsize];
  int i_style = FIX2INT(style);
  int i_length = FIX2INT(length);
  int i_line = FIX2INT(line);
  int i_filter = FIX2INT(filter);

  captcha(im, l, i_length, i_line, i_filter);
  makegif(im, gif, i_style);

  VALUE result = rb_ary_new2(2);
  rb_ary_push(result, rb_str_new2(l));
  rb_ary_push(result, rb_str_new(gif, gifsize));

  return result;
}
